atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/timestamp/TimestampServiceTests.java

```diff
@@ -0,0 +1,141 @@
+                        expectedNumTimestamps, actualNumTimestamps, range.getLowerBound(), range.getUpperBound()),
+                (int) actualNumTimestamps,
+                is(both(greaterThanOrEqualTo(1)).and(lessThanOrEqualTo(expectedNumTimestamps))));
+    }
+
+    public static void timestampRangesAreReturnedInNonOverlappingOrder(TimestampService timestampService) {
+        List<TimestampRange> timestampRanges = new ArrayList<>();
+
+        timestampRanges.add(timestampService.getFreshTimestamps(10));
+        timestampRanges.add(timestampService.getFreshTimestamps(10));
+
+        long firstUpperBound = timestampRanges.get(0).getUpperBound();
+        long secondLowerBound = timestampRanges.get(1).getLowerBound();
+
+        assertThat(firstUpperBound, is(lessThan(secondLowerBound)));
+    }
+
+    public static void canRequestMoreTimestampsThanAreAllocatedAtOnce(TimestampService timestampService) {
+        for (int i = 0; i < ONE_MILLION / 1000; i++) {
+            timestampService.getFreshTimestamps(1000);
+        }
+
+        assertThat(timestampService.getFreshTimestamp(), is(ONE_MILLION + 1));
+    }
+
+    public static void willNotHandOutTimestampsEarlierThanAFastForward(
+            TimestampMigrationService timestampMigrationService, TimestampService timestampService) {
+        timestampMigrationService.fastForwardTimestamp(TWO_MILLION);
+
+        assertThat(timestampService.getFreshTimestamp(), is(greaterThan(TWO_MILLION)));
+    }
+
+    public static void willDoNothingWhenFastForwardToEarlierTimestamp(
+            TimestampMigrationService timestampMigrationService,
+            TimestampService timestampService) {
+        timestampMigrationService.fastForwardTimestamp(TWO_MILLION);
+        long ts1 = timestampService.getFreshTimestamp();
+        timestampMigrationService.fastForwardTimestamp(ONE_MILLION);
+        long ts2 = timestampService.getFreshTimestamp();
+        assertThat(ts2, greaterThan(TWO_MILLION));
+        assertThat(ts2, greaterThan(ts1));
+    }
+
+    public static void canReturnManyUniqueTimestampsInParallel(TimestampService timestampService)
+            throws InterruptedException, TimeoutException {
+        Set<Long> uniqueTimestamps = new ConcurrentSkipListSet<>();
+
+        repeat(TWO_MILLION, () -> uniqueTimestamps.add(timestampService.getFreshTimestamp()));
+
+        assertThat(uniqueTimestamps.size(), is((int) TWO_MILLION));
+    }
+
+    private static void repeat(long count, Runnable task) throws InterruptedException, TimeoutException {
+        for (int i = 0; i < count; i++) {
+            executor.submit(task);
+        }
+
+        executor.shutdown();
+        executor.awaitTermination(1, TimeUnit.MINUTES);
+
+        if (!executor.isTerminated()) {
+            throw new TimeoutException("Timed out waiting for the executor to terminate");
+        }
+    }
+
+    public static void shouldThrowIfRequestingNegativeNumbersOfTimestamps(TimestampService timestampService) {
```

Comment:
As mentioned above, we can use `@Test(expected = IllegalArgumentException.class)` here, if we make this an abstract class.